lambda_builders/lib/base.py

```diff
@@ -0,0 +1,145 @@
+"""
+Implementation of a basic builder
+"""
+
+import os
+import logging
+import six
+
+from collections import namedtuple
+from lambda_builders.lib.registry import BuilderMetaClass
+from lambda_builders.exceptions import BuilderError, BuilderFailed
+
+LOG = logging.getLogger(__name__)
+
+
+class ActionError(Exception):
+    """
+    Base class for exceptions raised by Actions. Raise this exception when the action ran into an unexpected failure.
+    """
+    pass
+
+
+class ActionFailed(Exception):
+    """
+    Base class for exception raised when action failed to complete. Use this to express well-known failure scenarios.
+    """
+    # TODO: Do we need a `error_code` of some sorts here? If so, we should create a named list of error codes.
+    pass
+
+
+# Named tuple to express the capabilities supported by the builder.
+# ``Language`` is the programming language. Ex: Python
+# ``LangageFramework`` is the framework of particular language. Ex: PIP
+# ``ApplicationFramework`` is the specific application framework used to write the code. Ex: Chalice
+BuilderCapability = namedtuple('BuilderCapability', ["Language", "LanguageFramework", "ApplicationFramework"])
```

Comment:
An application could be written in Python + PIP + an app framework like Chalice/Zappa/Serverless Framework. So this becomes a whole new builder that knows how to handle Chalice+Python+PIP projects. It will simply shell out to `chalice package` and optionally do a few more things. My goal was to support building not just vanilla Python/JS/Java projects, but also projects that use specialized app frameworks that need a custom action